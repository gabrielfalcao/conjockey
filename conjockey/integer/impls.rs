use crate::integer::Integer;

impl std::ops::Add for Integer {
    type Output = Integer;
    fn add(self, rhs: Integer) -> Integer {
        Integer(self.as_i64() + rhs.as_i64())
    }
}

impl std::ops::Sub for Integer {
    type Output = Integer;
    fn sub(self, rhs: Integer) -> Integer {
        Integer(self.as_i64() - rhs.as_i64())
    }
}

impl std::ops::AddAssign for Integer {
    fn add_assign(&mut self, rhs: Integer) {
        self.0 += rhs.as_i64()
    }
}

impl std::ops::SubAssign for Integer {
    fn sub_assign(&mut self, rhs: Integer) {
        self.0 -= rhs.as_i64()
    }
}

impl std::ops::Mul for Integer {
    type Output = Integer;
    fn mul(self, rhs: Integer) -> Integer {
        Integer(self.as_i64() * rhs.as_i64())
    }
}

impl std::ops::Div for Integer {
    type Output = Integer;
    fn div(self, rhs: Integer) -> Integer {
        Integer(self.as_i64() / rhs.as_i64())
    }
}

impl std::ops::MulAssign for Integer {
    fn mul_assign(&mut self, rhs: Integer) {
        self.0 *= rhs.as_i64()
    }
}

impl std::ops::DivAssign for Integer {
    fn div_assign(&mut self, rhs: Integer) {
        self.0 /= rhs.as_i64()
    }
}

impl std::ops::Rem for Integer {
    type Output = Integer;
    fn rem(self, rhs: Integer) -> Integer {
        Integer(self.as_i64() % rhs.as_i64())
    }
}

impl std::ops::RemAssign for Integer {
    fn rem_assign(&mut self, rhs: Integer) {
        self.0 %= rhs.as_i64()
    }
}

impl std::ops::Neg for Integer {
    type Output = Integer;
    fn neg(self) -> Integer {
        Integer(-self.as_i64())
    }
}

impl std::ops::Not for Integer {
    type Output = Integer;
    fn not(self) -> Integer {
        Integer(!self.as_i64())
    }
}

impl std::cmp::PartialOrd for Integer {
    fn partial_cmp(&self, rhs: &Integer) -> Option<std::cmp::Ordering> {
        self.as_i64().partial_cmp(&rhs.as_i64())
    }
}

impl std::cmp::Ord for Integer {
    fn cmp(&self, rhs: &Integer) -> std::cmp::Ordering {
        self.as_i64().cmp(&rhs.as_i64())
    }
}

impl crate::traits::ops::algebraic::AlgebraicOperations for Integer {}

impl From<i64> for Integer {
    fn from(value: i64) -> Integer {
        Integer(value)
    }
}

impl From<i32> for Integer {
    fn from(value: i32) -> Integer {
        Integer(Into::<i64>::into(value))
    }
}

impl Integer {
    pub fn is_signed(self) -> bool {
        self.0 < 0
    }
    pub fn is_unsigned(self) -> bool {
        self.0 >= 0
    }
    pub fn as_signed(self) -> Option<i64> {
        if self.is_signed() {
            Some(self.0)
        } else {
            None
        }
    }
    pub fn as_unsigned(self) -> Option<u64> {
        self.as_unsigned_u64()
    }
    pub fn as_unsigned_u64(self) -> Option<u64> {
        match TryInto::<u64>::try_into(self.0) {
            Ok(value) => Some(value),
            Err(_) => None,
        }
    }
    pub fn as_unsigned_u32(self) -> Option<u32> {
        match TryInto::<u32>::try_into(self.0) {
            Ok(value) => Some(value),
            Err(_) => None,
        }
    }
    pub fn as_unsigned_u16(self) -> Option<u16> {
        match TryInto::<u16>::try_into(self.0) {
            Ok(value) => Some(value),
            Err(_) => None,
        }
    }
    pub fn as_unsigned_u8(self) -> Option<u8> {
        match TryInto::<u8>::try_into(self.0) {
            Ok(value) => Some(value),
            Err(_) => None,
        }
    }
    pub fn to_unsigned(self) -> u64 {
        self.to_unsigned_u64()
    }
    pub fn to_unsigned_u64(self) -> u64 {
        match self.as_unsigned_u64() {
            Some(value) => value,
            None => panic!("{} cannot be converted into u64", self.0),
        }
    }
    pub fn to_unsigned_u32(self) -> u32 {
        match self.as_unsigned_u32() {
            Some(value) => value,
            None => panic!("{} cannot be converted into u32", self.0),
        }
    }
    pub fn to_unsigned_u16(self) -> u16 {
        match self.as_unsigned_u16() {
            Some(value) => value,
            None => panic!("{} cannot be converted into u16", self.0),
        }
    }
    pub fn to_unsigned_u8(self) -> u8 {
        match self.as_unsigned_u8() {
            Some(value) => value,
            None => panic!("{} cannot be converted into u8", self.0),
        }
    }
}

impl std::ops::BitAnd for Integer {
    type Output = Integer;
    fn bitand(self, rhs: Integer) -> Integer {
        Integer(self.as_i64() & rhs.as_i64())
    }
}

impl std::ops::BitOr for Integer {
    type Output = Integer;
    fn bitor(self, rhs: Integer) -> Integer {
        Integer(self.as_i64() | rhs.as_i64())
    }
}

impl std::ops::BitAndAssign for Integer {
    fn bitand_assign(&mut self, rhs: Integer) {
        self.0 &= rhs.as_i64()
    }
}

impl std::ops::BitOrAssign for Integer {
    fn bitor_assign(&mut self, rhs: Integer) {
        self.0 |= rhs.as_i64()
    }
}

impl std::ops::BitXor for Integer {
    type Output = Integer;
    fn bitxor(self, rhs: Integer) -> Integer {
        Integer(self.as_i64() ^ rhs.as_i64())
    }
}

impl std::ops::BitXorAssign for Integer {
    fn bitxor_assign(&mut self, rhs: Integer) {
        self.0 ^= rhs.as_i64()
    }
}

impl std::ops::Shr for Integer {
    type Output = Integer;
    fn shr(self, rhs: Integer) -> Integer {
        Integer(self.as_i64() >> rhs.as_i64())
    }
}

impl std::ops::Shl for Integer {
    type Output = Integer;
    fn shl(self, rhs: Integer) -> Integer {
        Integer(self.as_i64() << rhs.as_i64())
    }
}

impl std::ops::ShrAssign for Integer {
    fn shr_assign(&mut self, rhs: Integer) {
        self.0 >>= rhs.as_i64()
    }
}

impl std::ops::ShlAssign for Integer {
    fn shl_assign(&mut self, rhs: Integer) {
        self.0 <<= rhs.as_i64()
    }
}

impl crate::traits::ops::LogicalOperations for Integer {}
