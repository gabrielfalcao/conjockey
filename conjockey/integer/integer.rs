#[derive(Debug, Copy, Clone, PartialEq, Eq)]
pub struct Integer(pub i64);

impl Integer {
    pub fn as_i64(self) -> i64 {
        self.0
    }
}
