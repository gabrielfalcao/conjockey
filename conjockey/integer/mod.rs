mod impls;
#[cfg(test)]
mod tests;
mod json;
mod yaml;

mod integer;
pub use integer::Integer;
