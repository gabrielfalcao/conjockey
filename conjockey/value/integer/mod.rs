mod impls;
#[cfg(test)]
mod tests;
