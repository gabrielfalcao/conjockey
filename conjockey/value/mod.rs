mod null;
mod boolean;
mod string;
mod json;
mod plist;
mod toml;
mod yaml;
mod value;
mod integer;

pub use value::*;
