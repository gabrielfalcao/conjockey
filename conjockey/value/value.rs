use crate::integer::Integer;

#[derive(Clone, Debug, Eq, PartialEq)]
pub enum Value {
    Null,
    Boolean(bool),
    String(String),
    Integer(Integer),
    // Float(Float),
    // DateTime(DateTime),
    // Array(Array),
    // Map(Map<String, Value>),
    Unknown,
}
