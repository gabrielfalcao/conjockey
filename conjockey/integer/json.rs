use crate::integer::Integer;
use crate::value::Value;
use serde_json::Number as JsonNumber;
use serde_json::Value as JsonValue;
use std::any::TypeId;

impl From<JsonNumber> for Integer {
    fn from(value: JsonNumber) -> Integer {
        match value.as_i64() {
            Some(value) => Integer(value),
            None => panic!(
                "can not convert {:#?} into i64 for conjockey::Integer",
                value
            ),
        }
    }
}
impl From<Integer> for JsonNumber {
    fn from(value: Integer) -> JsonNumber {
        match value.as_signed_i64() {
            Some(value) => JsonNumber::from(value),
            None => panic!(
                "can not convert {:#?} into i64 for {:#?}",
                value,
                TypeId::of::<JsonNumber>()
            ),
        }
    }
}

impl Into<JsonNumber> for Value {
    fn into(self) -> JsonNumber {
        match self {
            Value::Integer(value) => Into::<JsonNumber>::into(value),
            _ => panic!(
                "can not convert {:#?} to {:#?}",
                self,
                TypeId::of::<Value>()
            ),
        }
    }
}

impl Into<JsonValue> for Integer {
    fn into(self) -> JsonValue {
        JsonValue::Number(JsonNumber::from(self.to_signed_i64()))
    }
}
