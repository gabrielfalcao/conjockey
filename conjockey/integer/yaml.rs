use crate::integer::Integer;
use crate::value::Value;
use serde_yaml::Number as YamlNumber;
use serde_yaml::Value as YamlValue;
use std::any::TypeId;

impl From<YamlNumber> for Integer {
    fn from(value: YamlNumber) -> Integer {
        match value.as_i64() {
            Some(value) => Integer(value),
            None => panic!(
                "can not convert {:#?} into i64 for conjockey::Integer",
                value
            ),
        }
    }
}
impl From<Integer> for YamlNumber {
    fn from(value: Integer) -> YamlNumber {
        match value.as_signed_i64() {
            Some(value) => YamlNumber::from(value),
            None => panic!(
                "can not convert {:#?} into i64 for {:#?}",
                value,
                TypeId::of::<YamlNumber>()
            ),
        }
    }
}

impl Into<YamlNumber> for Value {
    fn into(self) -> YamlNumber {
        match self {
            Value::Integer(value) => Into::<YamlNumber>::into(value),
            _ => panic!(
                "can not convert {:#?} to {:#?}",
                self,
                TypeId::of::<Value>()
            ),
        }
    }
}

impl Into<YamlValue> for Integer {
    fn into(self) -> YamlValue {
        YamlValue::Number(YamlNumber::from(self.to_signed_i64()))
    }
}
