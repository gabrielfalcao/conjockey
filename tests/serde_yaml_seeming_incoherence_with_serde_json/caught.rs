use conjockey::integer::Integer;
use conjockey::value::Value;
use serde_yaml::Number as YamlNumber;
use serde_yaml::Value as YamlValue;

fn test_yaml_integer() {
    assert_eq!(
        YamlValue::Number(YamlNumber::from(501)),
        Into::<YamlValue>::into(Value::Integer(Integer(501))),
    );
    assert_eq!(
        Value::Integer(Integer(501)),
        Into::<Value>::into(YamlValue::Number(YamlNumber::from(501))),
    );
}

fn test_yaml_integer_from_option_integer() {
    assert_eq!(
        YamlValue::Number(YamlNumber::from(501)),
        Into::<YamlValue>::into(Some(Integer(501))),
    );
}

fn test_yaml_integer_from_option_value_integer() {
    assert_eq!(
        YamlValue::Number(YamlNumber::from(501)),
        Into::<YamlValue>::into(Some(Value::Integer(Integer(501)))),
    );
}

#[test]
fn test() {
    test_yaml_integer();
    test_yaml_integer_from_option_integer();
    test_yaml_integer_from_option_value_integer();
}

fn main() {
    test_yaml_integer();
    test_yaml_integer_from_option_integer();
    test_yaml_integer_from_option_value_integer();
}
