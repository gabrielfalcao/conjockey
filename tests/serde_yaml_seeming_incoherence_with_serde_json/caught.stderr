error[E0277]: the trait bound `serde_yaml::Value: From<Option<conjockey::Integer>>` is not satisfied
  --> tests/serde_yaml_seeming_incoherence_with_serde_json/caught.rs:20:33
   |
20 |         Into::<YamlValue>::into(Some(Integer(501))),
   |         ----------------------- ^^^^^^^^^^^^^^^^^^ the trait `From<Option<conjockey::Integer>>` is not implemented for `serde_yaml::Value`
   |         |
   |         required by a bound introduced by this call
   |
   = help: the following other types implement trait `From<T>`:
             `serde_yaml::Value` implements `From<&[T]>`
             `serde_yaml::Value` implements `From<&str>`
             `serde_yaml::Value` implements `From<Cow<'_, str>>`
             `serde_yaml::Value` implements `From<Vec<T>>`
             `serde_yaml::Value` implements `From<bool>`
             `serde_yaml::Value` implements `From<f32>`
             `serde_yaml::Value` implements `From<f64>`
             `serde_yaml::Value` implements `From<i16>`
           and $N others
   = note: required for `Option<conjockey::Integer>` to implement `Into<serde_yaml::Value>`

error[E0277]: the trait bound `serde_yaml::Value: From<Option<conjockey::Value>>` is not satisfied
  --> tests/serde_yaml_seeming_incoherence_with_serde_json/caught.rs:27:33
   |
27 |         Into::<YamlValue>::into(Some(Value::Integer(Integer(501)))),
   |         ----------------------- ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `From<Option<conjockey::Value>>` is not implemented for `serde_yaml::Value`
   |         |
   |         required by a bound introduced by this call
   |
   = help: the following other types implement trait `From<T>`:
             `serde_yaml::Value` implements `From<&[T]>`
             `serde_yaml::Value` implements `From<&str>`
             `serde_yaml::Value` implements `From<Cow<'_, str>>`
             `serde_yaml::Value` implements `From<Vec<T>>`
             `serde_yaml::Value` implements `From<bool>`
             `serde_yaml::Value` implements `From<f32>`
             `serde_yaml::Value` implements `From<f64>`
             `serde_yaml::Value` implements `From<i16>`
           and $N others
   = note: required for `Option<conjockey::Value>` to implement `Into<serde_yaml::Value>`
